### ISSUE TITLE
PR #19 scan path: ptxas compile pathology on 22-tetrode 2D HPC shape

### ISSUE THREAD

**Issue by edeno** [LorenFrankLab/non_local_detector#21]
## Summary

PR #19's scan-path (`_predict_nonlocal_electrode_scan_impl` added in commit `f248064`) produces a fused XLA kernel that `ptxas` cannot compile in bounded time on common production shapes.

## Reproducer

- Session: `j1620210710_.nwb` / epoch `02_r1` (spatial-bandit; HPC CA1, 22 tetrodes, 4-D waveform features, ~418k total spikes, 709k time bins at 0.5 ms)
- Environment: 2D 3.5 cm grid (3420 position bins)
- Detector: `ContFragClusterlessClassifier` (2-state, ContFrag)
- Algorithm: `clusterless_algorithm="clusterless_kde_log"`
- Fresh JAX process, A100 80 GB, `jax_enable_x64=True`

```python
detector = ContFragClusterlessClassifier(
    environments=env,
    clusterless_algorithm="clusterless_kde_log",
    clusterless_algorithm_params={"position_std": ..., "waveform_std": ..., "block_size": 1000},
)
detector.fit(position_time=..., position=..., spike_times=..., spike_waveform_features=...)
detector.predict(...)  # hangs in ptxas >30 min, never completes
```

`ptxas` runs at 99% CPU with `--warn-on-spills`, indicating register-allocation failure on the fused scan body. XLA emits:
```
[Compiling module input_reduce_fusion.1 for GPU] Very slow compile?
```

## A/B comparison

| Config | compile+first | steady (min/med/max) | status |
|---|---|---|---|
| post-PR `kde` (default, untouched by #19) | 138 s | 83 / 85 / 88 s | ✅ works |
| pre-PR `kde_log` (`main` at `7e2d1d9`, Python for-loop over electrodes) | 168 s | **83 / 83 / 85 s** | ✅ works |
| **post-PR `kde_log`** (`main` at `fcf5100`, scan path) | **>30 min, killed** | N/A | ❌ **regression** |

`ptxas` was still busy at 30 min when I killed it — the compile may eventually complete, but it's not bounded in any useful sense for interactive work.

## Why pre-merge validation missed this

The pre-merge real-data validation for PR #19 (see comparison files at `/tmp/pr-validation/`) ran with the default `clusterless_algorithm="clusterless_kde"` — which maps to `clusterless_kde.py`, a module PR #19 did not modify. The scan path lives in `clusterless_kde_log.py`, and the validation never exercised it end-to-end on real data. Unit tests and the synthetic 64-tetrode benchmark (where ptxas apparently does manage to compile the kernel) passed.

## Cross-path correctness

Comparing the posteriors from the two working paths on the same data:

| var | max\|diff\| | p99 | median | rel_max |
|---|---|---|---|---|
| `acausal_posterior` (709k × 6840) | 8.1e-06 | 6.0e-08 | 1.6e-12 | 8.5e-06 |
| `acausal_state_probabilities` | 1.1e-05 | 2.7e-06 | 2.8e-08 | 1.1e-05 |

`kde` (default, product-space) and pre-PR `kde_log` (log-space Python loop) agree to ~1e-5 at fp64 — well within `jnp.exp(log_x) ≈ x` noise. Both modules are numerically correct. This is a pure compile-time regression, not a correctness regression.

## Suspected root cause

`_predict_nonlocal_electrode_scan_jit` (PR #19 commit `f248064`) uses `jax.lax.scan` over size-bucketed electrode batches. Inside the scan body is the fused `block_estimate_log_joint_mark_intensity` + `segment_sum` (commits `aec22af`, `a7edf6d`) + compensated-linear-marginal. XLA's `input_reduce_fusion` pass fuses this into a single kernel with high register pressure; on 4-D waveforms × 3420 position bins × a batch of ~20 electrodes × a block size of 1000 decoding spikes, ptxas spills and runs an expensive search that doesn't converge in practice.

## Options

1. **Gate the scan path** behind an opt-in flag (default off on this kind of shape) with automatic fallback to the per-electrode Python loop if compile exceeds a timeout.
2. **Per-bucket pre-check**: skip the scan path for buckets with estimated compiled-kernel size above a threshold, falling back to per-electrode compile (which works — pre-PR path is fine).
3. **Tune XLA flags** (`--xla_gpu_enable_triton_gemm=false`, `--xla_gpu_disable_async_collectives`, etc.) and see if ptxas can get through. Risky: changes behavior for other callers.
4. **Manual shape/tile adjustment**: force smaller scan windows (split the 22 electrodes into multiple smaller scan invocations) so each fused kernel is smaller. Loses the Python dispatch reduction but avoids the compile blow-up.
5. **Revert #19** and rework the PR with a shape-aware scan-vs-loop router baked in.

## Reproduction artifacts

Available in `/tmp/pr19-postmerge-validation/` on `zephyr`:

- `main-post_summary.json` (post-PR kde working; kde_log killed)
- `main-pre_summary.json` (pre-PR kde_log working)
- `main-post_2d_contfrag_clusterless_kde_clusterless_results.nc`
- `main-pre_2d_contfrag_clusterless_kde_log_clusterless_results.nc`
- `main-pre-run.log`, `main-post-run.log`

## Related

- PR #19 `fcf5100` — introduced scan path
- PR #14 (`benchmark/kde-log-vs-product`, open since April 1) — modifies `kde_distance` on the default path; would help numerical stability on default (non-log) path

**Comment by edeno** [comment]
## Post-merge validation update — confirmed scope

Ran the investigation-branch validation today:

### Validated: Tasks 1/2/4 + Phase 0 are correct on real data

Forced the fallback Python-loop-per-electrode path with `enc_tile_size=4096` on `gpu-opt-investigation` branch (= `fcf5100`, the reverted PR #19 state). Compared against pre-PR `clusterless_kde_log` output on the same data:

| var | max\|diff\| | p99 | median | rel_max |
|---|---|---|---|---|
| `acausal_posterior` (709k × 6840) | 8.1e-06 | 6.0e-08 | 1.6e-12 | 8.5e-06 |
| `acausal_state_probabilities` | 1.1e-05 | 2.7e-06 | 2.8e-08 | 1.1e-05 |

Within `jnp.exp(log_x) ≈ x` ULP noise at fp64.  **Tasks 1/2/4 + Phase 0 are numerically correct on 2D 22-tetrode HPC real-data shape.** The only broken piece is Task 3 (scan).

### Validated: fallback compiles (scan doesn't)

| Config | compile+first | steady median |
|---|---|---|
| post-PR fallback (`enc_tile_size=4096`) | **301.5 s** (completes) | 177 s |
| post-PR scan (`enc_tile_size=None`) | **>30 min stuck**, killed | N/A |
| pre-PR fallback (`enc_tile_size=None`, no scan) | 168 s | 83 s |

So the pathology is specifically the **scan-over-electrodes fused kernel** (`_predict_nonlocal_electrode_scan_impl` from commit `f248064`), not the broader set of fori_loop / GEMM-hoist / segment_sum changes.

### Resolution: PR #19 reverted via #22

Per post-mortem analysis (see #22 comment), the compile pathology + the fact that Tasks 1/2/4 + Phase 0 only deliver ~5-10% runtime win at typical shapes without scan means a surgical cherry-pick isn't worth the code cost.  A cleaner redesign is the right path.

### Requirements for a redesigning PR (target of this issue)

1. **Shape-aware scan-vs-loop router** — compile-timeout fallback baked in from day one, not a post-hoc flag.
2. **Real-data validation on ≥3 distinct shapes** before merge — at minimum: 1D HPC + 2D HPC + one high-tetrode-count synthetic. Run with the specific `clusterless_algorithm` the PR optimizes, not just the default.
3. **Compile-time as a pass/fail criterion** alongside wall-clock and accuracy. Flag any compile >20 min on a production-like shape as a regression.
4. **Cross-path parity check** (`kde` vs `kde_log`) in the PR's validation artifacts to confirm numerical correctness without relying on either module as "the reference".

### Investigation branch preserved

`gpu-opt-investigation` retains the full original PR #19 commit history for reference / future cherry-picking of specific commits if needed.

### Related

- Revert: #22
- Validation workflow memory: see `.claude/` memory files (updated post-PR-19)
